### PR TITLE
[Yoga Beta] Improvements to the experimental support for Yoga layout.

### PR DESCRIPTION
**Pull request by appleguy**
Yoga remains an unsupported / speculative feature, but these improvements are important for
the functionality of clients that are experimenting with it.

Total list of changes:

1. Add "YGDirection" field to control vertical and horizontal.
2. Add .start, .end, etc to gated EdgeInsets struct to support RTL layout properly. I can rename this struct to be Yoga-specific if desired, but it does need to exist on ASLayoutElementStyle objects.
3. Fix some of the logic in the Yoga-private layout extensions. ASDisplayNode+Yoga is compiled out of most builds, so it is not possible for these changes to affect mainline users.

For example, without these changes, ASButtonNode is not able to lay out correctly. These
changes allow certain subtrees that use layout specs to coexist properly in a Yoga heirarchy.

The most significant change here is moving ASEdgeInsets into the #if YOGA gating. Although
this is technically an API change, this type was added with no known use cases and is
really only useful for flexbox layout specification. So, before usages of it are created,
it makes sense to constrain the Texture API surface until that time.

### CODE REVIEW

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
Build failure:

**Comment by appleguy** [comment]
This has now merged latest master and should be ready to land, including RTL support as provided by PR #60.

**Comment by pinterest-ios** [comment]
Build failure:

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/7caaf710cbdd713669d111c93a306e7e/raw/c98beaaf0d534f447b01623afddf52fe9436b86d/log.txt)

File: Source/Layout/ASLayoutElement.mm [hunk 1]
```diff
@@ -150,7 +150,8 @@ @implementation ASLayoutElementStyle {
   std::atomic<CGPoint> _layoutPosition;
 
 #if YOGA
-  std::atomic<ASStackLayoutDirection> _direction;
+  std::atomic<ASStackLayoutDirection> _flexDirection;
+  std::atomic<YGDirection> _direction;
```
Comment (Adlai-Holler) [inline Source/Layout/ASLayoutElement.mm]: (You knew it was coming!) Maybe these would be better-off as Objective-C atomic properties to improve performance and cut down on boilerplate?

File: Source/ASDisplayNode+Yoga.mm [hunk 2]
```diff
@@ -333,8 +337,24 @@ - (void)invalidateCalculatedYogaLayout
...
 }
 
+- (void)semanticContentAttributeDidChange:(UISemanticContentAttribute)attribute
+{
+  if (AS_AT_LEAST_IOS9) {
+    UIUserInterfaceLayoutDirection layoutDirection =
+                         [UIView userInterfaceLayoutDirectionForSemanticContentAttribute:attribute];
+    self.style.direction = (layoutDirection == UIUserInterfaceLayoutDirectionLeftToRight
+                                             ? YGDirectionLTR : YGDirectionRTL);
+  }
```
Comment (Adlai-Holler) [inline Source/ASDisplayNode+Yoga.mm]: Nice! That class method may not be safe off-main, but I expect you've considered that. Even if so, the risk here is teeny tiny.

File: Source/ASDisplayNode+Beta.h [hunk 3]
```diff
@@ -176,12 +176,14 @@ extern void ASDisplayNodePerformBlockOnEveryYogaChild(ASDisplayNode * _Nullable
...
 - (void)calculateLayoutFromYogaRoot:(ASSizeRange)rootConstrainedSize;
+- (void)semanticContentAttributeDidChange:(UISemanticContentAttribute)attribute;
 
 @end
 
 @interface ASLayoutElementStyle (Yoga)
 
-@property (nonatomic, assign, readwrite) ASStackLayoutDirection direction;
+@property (nonatomic, assign, readwrite) ASStackLayoutDirection flexDirection;
+@property (nonatomic, assign, readwrite) YGDirection direction;
```
Comment (Adlai-Holler) [inline Source/ASDisplayNode+Beta.h]: Even if these are implemented using C++ atomics, we ought to declare them as `atomic`.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by maicki** [comment]
As I'm currently working in this area and this PR can introduce some merge conflict I rather would prevent I will try to get this PR further and hopefully in sooner than later cc @appleguy @Adlai-Holler

File: Source/Layout/ASLayoutElement.mm (hunk 1 above)
Comment (maicki) [inline Source/Layout/ASLayoutElement.mm]: As all of them are currently c++ atomics let's go with it for now and do the whole change in a another diff if we want to.

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/f00ea634c8f35ce1ffa4f3fc353bc139/raw/aa504c8eef5b00271ae5b9d05c9449272c13e6bf/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/ef06ae6bc1d32b2706014b517fc14bee/raw/d1eccf3a7e32886ee08306ec10b1ac7e7e929f9f/log.txt)

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/a2e7abd305d7eb52f82d93df670dec31/raw/23706fb41dbaeca9a985a984bc2b6e32be06a377/log.txt)

**Comment by appleguy** [comment]
@maicki @Adlai-Holler thanks for your comments here and for merging this - sorry for the slow response on my side. I could put up another change that includes atomic, too, but won't worry about it until the other changes @maicki mentioned are landed.

One more thing - @maicki it's totally OK if your refactor removes the Yoga hook from ASDisplayNode entirely. Just drop me a message if you do this and I will work on re-adding it in the least invasive way possible. I actually had some recent ideas about how to possibly use the ASLayoutSpec API to build the Yoga tree; it would be a bit unusual because it would have to connect with the Yoga tree across levels of the hierarchy, BUT the integration points would basically not touch ASDisplayNode.

**Pull request merged**